File: src/vibe_kanban_clone/api/routes/projects.py
"""Project routes."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from vibe_kanban_clone.api.deps import get_session
from vibe_kanban_clone.api.routes.ws import broadcast_global
from vibe_kanban_clone.schemas.common import PaginatedParams
from vibe_kanban_clone.schemas.project import ProjectCreate, ProjectRead, ProjectUpdate
from vibe_kanban_clone.services import projects as projects_service

router = APIRouter()


@router.get("/projects", response_model=list[ProjectRead])
async def list_projects(
    pagination: Annotated[PaginatedParams, Depends()],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[ProjectRead]:
    """List all projects."""
    return await projects_service.list_projects(session, pagination.limit, pagination.offset)


@router.post("/projects", response_model=ProjectRead, status_code=201)
async def create_project(
    data: ProjectCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> ProjectRead:
    """Create a new project."""
    project = await projects_service.create_project(session, data)
    await broadcast_global(
        "project.created",
        ProjectRead.model_validate(project).model_dump(mode="json"),
    )
    return project


@router.get("/projects/{project_id}", response_model=ProjectRead)
async def get_project(
    project_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> ProjectRead:
    """Get a project by ID."""
    project = await projects_service.get_project(session, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


@router.patch("/projects/{project_id}", response_model=ProjectRead)
async def update_project(
    project_id: int,
    data: ProjectUpdate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> ProjectRead:
    """Update a project."""
    project = await projects_service.get_project(session, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    project = await projects_service.update_project(session, project, data)
    await broadcast_global(
        "project.updated",
        ProjectRead.model_validate(project).model_dump(mode="json"),
    )
    return project


@router.delete("/projects/{project_id}", status_code=204)
async def delete_project(
    project_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    """Delete a project."""
    project = await projects_service.get_project(session, project_id)
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    await projects_service.delete_project(session, project)
    await broadcast_global(
        "project.deleted",
        {"project_id": project_id},
    )

File: server/src/routes/projects.ts
import { zValidator } from "@hono/zod-validator";
import { Hono } from "hono";
import { HTTPException } from "hono/http-exception";
import { ProjectCreate, ProjectUpdate } from "../schemas/project.js";
import { PaginatedParams } from "../schemas/common.js";
import * as projectsService from "../services/projects.js";
import { broadcastGlobal } from "../ws/manager.js";

const app = new Hono();

app.get("/projects", zValidator("query", PaginatedParams), async (c) => {
  const { limit, offset } = c.req.valid("query");
  const projects = await projectsService.listProjects(limit, offset);
  return c.json(projects);
});

app.post("/projects", zValidator("json", ProjectCreate), async (c) => {
  const data = c.req.valid("json");
  const project = await projectsService.createProject(data);
  await broadcastGlobal("project.created", project);
  return c.json(project, 201);
});

app.get("/projects/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const project = await projectsService.getProject(id);
  if (!project) throw new HTTPException(404, { message: "Project not found" });
  return c.json(project);
});

app.patch("/projects/:id", zValidator("json", ProjectUpdate), async (c) => {
  const id = Number(c.req.param("id"));
  const data = c.req.valid("json");
  const project = await projectsService.updateProject(id, data);
  if (!project) throw new HTTPException(404, { message: "Project not found" });
  await broadcastGlobal("project.updated", project);
  return c.json(project);
});

app.delete("/projects/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const project = await projectsService.getProject(id);
  if (!project) throw new HTTPException(404, { message: "Project not found" });
  await projectsService.deleteProject(id);
  await broadcastGlobal("project.deleted", { project_id: id });
  return c.body(null, 204);
});

export default app;
